### PR TITLE
Defend HelpCompletion from null message from VSCode

### LINKED ISSUE #1230: [Error] TypeError: Cannot read property 'text' of undefined at HelpCompletionFeature.onEvent
### System Details

- Operating system name and version: `Windows 10 Enterprise 1709 [10.0.16299.98]`
- VS Code version: `1.21.0`
- PowerShell extension version: `1.6.0`
- Output from `$PSVersionTable`:
```
Name                           Value
----                           -----
PSVersion                      5.1.16299.98
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.16299.98
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

```
code -v
1.21.0
9a199d77c82fcb82f39c68bb33c614af01c111ba
x64

$pseditor.EditorServicesVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
1      6      0      0


code --list-extensions --show-versions
AlanWalk.markdown-toc@1.5.6
bierner.markdown-preview-github-styles@0.1.0
christian-kohler.path-intellisense@1.4.2
DavidAnson.vscode-markdownlint@0.14.0
dbaeumer.vscode-eslint@1.4.7
DotJoshJohnson.xml@1.9.2
ecmel.vscode-html-css@0.2.0
formulahendry.auto-close-tag@0.5.6
formulahendry.auto-rename-tag@0.0.15
formulahendry.code-runner@0.9.2
HookyQR.beautify@1.3.0
KnisterPeter.vscode-github@0.27.1
mdickin.markdown-shortcuts@0.8.1
mkaufman.HTMLHint@0.4.0
mrmlnc.vscode-remark@1.1.0
ms-vscode.csharp@1.14.0
ms-vscode.github-issues-prs@0.9.0
ms-vscode.PowerShell@1.6.0
ms-vsts.team@1.122.0
msjsdiag.debugger-for-chrome@4.2.0
robertohuertasm.vscode-icons@7.21.0
Shan.code-settings-sync@2.9.0
shinnn.stylelint@0.34.2
shuworks.vscode-table-formatter@1.2.1
steve8708.Align@0.2.0
streetsidesoftware.code-spell-checker@1.6.5
streetsidesoftware.code-spell-checker-german@0.1.2
Tyriar.sort-lines@1.5.0
Zignd.html-css-class-completion@1.17.1

$PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.16299.98
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.16299.98
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

### Issue Description

When using command completion an error gets thrown inside `Log (Window)`:
```
[2018-03-16 00:01:17.448] [renderer1] [error] Cannot read property 'text' of undefined: TypeError: Cannot read property 'text' of undefined
    at HelpCompletionFeature.onEvent (C:\Users\Velocet\.vscode\extensions\ms-vscode.powershell-1.6.0\src\features\HelpCompletion.ts:49:42)
    at e.fire (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:88:1012)
    at e.$acceptDirtyStateChanged (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:627:178)
    at e._doInvokeHandler (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:587:832)
    at e._invokeHandler (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:587:550)
    at e._receiveRequest (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:586:658)
    at e._receiveOneMessage (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:586:427)
    at c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:585:494
    at c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:588:343
    at c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:87:749
    at e.fire (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:88:992)
    at Socket.<anonymous> (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\node\extensionHostProcess.js:143:227)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at readableAddChunk (_stream_readable.js:178:18)
    at Socket.Readable.push (_stream_readable.js:136:10)
    at Pipe.onread (net.js:560:20)
```
Example: Writing `Write-Ho` throws an error when pressing Tab to complete the command.

Since i dunno much about the inner workings of VS Code or TypeScript, this is just a guess: The path referenced in the error (`C:\Users\Velocet\.vscode\extensions\ms-vscode.powershell-1.6.0\src\features\HelpCompletion.ts`) is wrong. The file in questsion is located at `C:\Users\Velocet\.vscode\extensions\ms-vscode.powershell-1.6.0\out\src\features\HelpCompletion.js` and is a JavaScript file instead of a TypeScript one.

### Attached Logs

[1521154104-9ebabe1c-9422-485b-954d-4e8dacc678161521154100855.zip](https://github.com/PowerShell/vscode-powershell/files/1817369/1521154104-9ebabe1c-9422-485b-954d-4e8dacc678161521154100855.zip)

### PR DESCRIPTION
**Pull request by rjmholt**
Small change to ensure that null messages from VSCode don't cause errors, to fix #1230.

### CODE REVIEW

**Comment by rjmholt** [comment]
So in addition to #1230, there is also #1189 and #1043. Do we think they are the same bug?

**Comment by rkeithhill** [comment]
It certainly looks that way.  My only concern is if the dev tools console err msg is just a harmless coincidence instead of what is actually causing the Intellisense problem in #1189.

**Comment by tylerl0706** [comment]
> I tried to install Wireshark, but apparently there's extra stuff you need to install on Windows if you want to listen to localhost/loopback...

@rjmholt maybe fiddler can do this?

File: src/features/HelpCompletion.ts [hunk 1]
```diff
@@ -44,6 +44,10 @@ export class HelpCompletionFeature implements IFeature {
     }
 
     public onEvent(changeEvent: TextDocumentChangeEvent): void {
+        if (!changeEvent) {
+            return;
+        }
+
         this.helpCompletionProvider.updateState(
             changeEvent.document,
             changeEvent.contentChanges[0].text,
```
Comment (rkeithhill) [inline src/features/HelpCompletion.ts]: Don't forget to undefined check `contentChanges[0]`.  If changeEvent itself was getting thru as undefined then we should have gotten an on `changeEvent.document` first.

**Comment by rkeithhill** [comment]
So I'm running this change and I see this log message during debugging of the extension:
```
typeDefinition.js:18
Bad change event message: {"document":{"uri":{"$mid":1,"fsPath":"c:\\Users\\Keith\\Documents\\WindowsPowerShell\\Profile.ps1","external":"file:///c%3A/Users/Keith/Documents/WindowsPowerShell/Profile.ps1","path":"/C:/Users/Keith/Documents/WindowsPowerShell/Profile.ps1","scheme":"file"},"fileName":"c:\\Users\\Keith\\Documents\\WindowsPowerShell\\Profile.ps1","isUntitled":false,"languageId":"powershell","version":2,"isClosed":false,"isDirty":true,"eol":2,"lineCount":31},"contentChanges":[]}
```
In looking at the docs I see this explanation of the onDidChangeTextDocument() method:
> An event that is emitted when a text document is changed. This usually happens when the contents changes but also when other things like the dirty-state changes.

I get this when I add a space to the end of a line but because I have VSCode set to trim trailing whitespace, I get this event but the contentChanges array is empty (0 length).  So I think this will generate incorrect "bad change" messages.  Here is what I recommend we change this to:
```posh
    public onEvent(changeEvent: TextDocumentChangeEvent): void {
        if (!(changeEvent && changeEvent.contentChanges)) {
            this.log.write(`Bad TextDocumentChangeEvent message: ${JSON.stringify(changeEvent)}`);
            return;
        }

        if (changeEvent.contentChanges.length > 0) {
            this.helpCompletionProvider.updateState(
                changeEvent.document,
                changeEvent.contentChanges[0].text,
                changeEvent.contentChanges[0].range);

            // Todo raise an event when trigger is found, and attach complete() to the event.
            if (this.helpCompletionProvider.triggerFound) {
                this.helpCompletionProvider.complete().then(() => this.helpCompletionProvider.reset());
            }
        }
    }
```

**Comment by rjmholt** [comment]
Aha! Man what an edge case! Yeah, I will implement that 😄  Thanks @rkeithhill!

**Comment by rkeithhill** [comment]
Ya know, given this TypeScript def of TextDocumentChangeEvent:

```
	export interface TextDocumentChangeEvent {

		/**
		 * The affected document.
		 */
		document: TextDocument;

		/**
		 * An array of content changes.
		 */
		contentChanges: TextDocumentContentChangeEvent[];
	}
```
Doesn't this mean those two fields can't be null or undefined?  Admittedly my TypeScript skillz are still pretty weak so I'm not 100% sure on that.  If it is true, we might as well just get rid of that first null check since it should never be invoked.

**Comment by tylerl0706** [comment]
@rkeithhill good find on the edge case! FWIW, I'm pretty sure and empty array is falsey so `changeEvent.contentChanges.length > 0` doesn't need the length check but it's easier to read so I'm cool with that. Just thought I'd point that out.

Also, I'm pretty sure that those two fields can still be null just like in a C# class. I don't think they can be undefined. I'm like 85% confident in this statement.

**Comment by rkeithhill** [comment]
After a quick experiment, it looks like you're right so leave the null check in.  But in the future, C# 8 won't allow null assignment unless the type is defined as nullable.  I'm looking forward to this. :-)

**Comment by rjmholt** [comment]
Yeah, [TypeScript's interface explanation page](https://www.typescriptlang.org/docs/handbook/interfaces.html) isn't very helpful. I think it's a compile-time check, which doesn't help much from arbitrary deserialised network messages, just gives completions on them. (I think the intent is that you use an interface to describe the JSON you're expecting, then you feed it into a constructor of a full TypeScript object to rehydrate it and give it methods, etc.).

Had no idea C# was planning that — finally we won't have to do null checks all the way down, and hopefully the CLR will be able to use the information to run quicker.

**Comment by rkeithhill** [comment]
Check out this MSDN article on [C# 8 - Nullable Reference Types](https://msdn.microsoft.com/en-us/magazine/mt829270.aspx).

**Comment by rkeithhill** [comment]
Time to merge this one.

**Pull request merged**